doc/man-source/mode2.1
.TH MODE2 "1" "Last change: Jan 2017 "mode2 @version@" "User Commands"
.SH NAME
.P
\fBmode2\fR - Show kernel output available for LIRC drivers

.SH SYNOPSIS
.P
\fBmode2\fR --driver <\fIdriver\fR> [--device <\fIdevice\fR>]
[\fIother options\fR]
.P
\fBmode2\fR --raw --device <\fIdevice\fR> [\fIother options\fR]
.P
\fBmode2\fR --driver <\fIdriver\fR>  --list-devices

.SH DESCRIPTION
.P
The original purpose of this program has been to check operation of
home-brew LIRC receiver hardware and to see the IR waveform of the
remote controller without an expensive oscilloscope.
Over time, \fBmode2\fR has evolved to a generic tool to check what
kind of output the kernel makes available for lirc.
.P
mode2 will simply print pulse and space lengths to stdout for drivers
supporting LIRC_MODE_MODE2. For drivers supporting LIRC_MODE_LIRCCODE such
as TV-cards and the irman driver writes the decoded, integer values.
.P
From version 0.9.4, \fBmode2\fR can also list the devices available for a
given driver. This requires enumeration support which is available
in most but not all drivers.
.P
When given the correct driver mode2 should always work. Without a driver,
the --raw option should always work for /dev/lirc* devices. At the time
of writing, usb devices also works in most cases. However, since it is
tricky to get the path to the usb device, such ones are normally best
accessed using a driver.
.P
Not all drivers requires a device.
.P

.SH OPTIONS

.TP
\fB\-d\fR \fB\-\-device\fR=\fIdevice\fR
Read from given device.
.TP
\fB\-H\fR \fB\-\-driver\fR=\fIdriver\fR
Use given driver.
.TP
\fB\-m\fR \fB\-\-mode\fR
Write pulse/space data as raw config files. Does not affect
LIRCCODE drivers.
.TP
\fB\-s\fR \fB\-\-scope\fR=\fItime\fR
Enable 'scope like display with time us per char. Does not
affect lirccode drivers.
.TP
\fB\-k\fR \fB\-\-keep-root\fR
Don't drop root privileges after opening device. See RUNNING AS ROOT.
.TP
\fB\-l\fR \fB\-\-list\-devices\fR
List all available devices for given driver. Requires support not
present in all drivers.
.TP
\fB\-g\fR \fB\-\-gap\fR=\fItime\fR
Treat spaces longer than time as the gap. Time is in microseconds.
.TP
\fB\-r\fR \fB\-\-raw\fR
Access device directly. This is primarely intended to be used with kernel rc
drivers i .e., when accessing \fI/dev/lirc*\fR devices. Using other devices
might work, but is in general undefined.
.TP
\fB\-A\fR \fB\-\-driver\-options\fR=\fIkey:value[|key:value...]\fR
Set driver-specific option named key to given value.
.TP
\fB\-U\fR \fB\-\-plugindir\fR=\fIdirectory\fR
Load drivers from directory. See DRIVER LOADING.
.TP
\fB\-O, \fB\-\-options-file\fR <\fIpath\fR>
File containing default values for all options. A relative path is
interpreted from current directory. See [FILES] below.
.TP
\fB\-D\fR \fB\-\-loglevel\fR=\fIlevel\fR
Determine the amount of logging information. [level] can be a symbolic
syslog level: 'error','warning, 'info', 'notice' or  'debug'. lirc
also defines three additional levels 'trace', 'trace1' and 'trace2' which
gives even more messages ('trace2' bringing the most). However, in the
log these messages are marked as 'debug'.
The level can also be an integer in the range 3 (almost no messages) to
10.
.TP
\fB\-h\fR \fB\-\-help\fR
Display usage summary.
.TP
\fB\-v\fR \fB\-\-version\fR
Display version.

.SH RUNNING AS ROOT
In many cases mode2 needs to run as root to access devices not available
to regular users. On the other hand, running as root creates problems
such as log files owned by root, security concerns etc.
.P
In order to cope with this, mode2 by default drops root privileges
after opening the input device. This support is based on that root
permissions are accquired using sudo(1) e. g., using
.nf

        $ sudo mode2 --raw /dev/lirc0

.fi
If not using sudo, the same behaviour could be accomplished using the
SUDO_USER environment variable e. g.,
.nf

        # SUDO_USER=$LOGNAME mode2 --raw /dev/lirc0

.fi
The --keep-root option will make mode2 to keep root privileges for the
complete run.

.SH EXAMPLES

Invoked with a MODE2 type driver with \fBmode2\fR prints space/mark timings:
.nf
     al@snorken $ ./mode2 --driver default --device /dev/lirc0
     Using driver default on device /dev/lirc0
     Trying device: /dev/lirc0
     Using device: /dev/lirc0
     pulse 2750
     space 800
     pulse 500
     space 350
     pulse 550
     space 350
     pulse 550
     ...
.if

.P

Invoked with a LIRCCODE type driver \fBmode2\fR prints decoded values:

.nf
     $ ./mode2 --driver atilibusb
     Using driver atilibusb on device /dev/lirc0
     Trying device: /dev/bus/usb/001/007
     Using device: /dev/bus/usb/001/007
     code: 0x14709b0000
     code: 0x14709b0000
     code: 0x14709b0000
     code: 0x14709b0000
     code: 0x14709b0000
     ...
.if

.P

Invoked using \fBmode2 --list-devices\fR for the devinput driver with a
single X10 RF usb device plugged in:
.nf
    $ ./mode2 --list-devices --driver devinput
    /dev/input/event12 [0bc7:0006] X10 WTI RF receiver vers:  1.10 serial: ?
    /dev/input/by-id/usb-X10_WTI_RF_receiver-event-if00 -> ../event12
    /dev/input/by-path/pci-0000:0b:00.0-usb-0:1:1.0-event -> ../event12
.if

.P

.SH "DRIVER LOADING"
Drivers are loaded dynamically. The search path used for this is determined
by (falling priority):
.IP \- 2
The --plugindir option.
.IP \- 2
The 'plugindir' entry in  the [lircd] section of the lirc_options.conf file.
.IP \- 2
The environment variable LIRC_PLUGINDIR.
.IP \- 2
A hardcoded default (@libpath@/lirc/plugins).

.SH ENVIRONMENT
.TP 4
.B LIRC_LOGLEVEL
Used as fallback when there is no --loglevel option present. Syntax
is the same as for --loglevel, see OPTIONS.

.SH "FILES"
.TP 4
.B @etcdir@/lirc/lirc_options.conf
The options file holding default values for command line options in
the [mode2] section. For some values including debug, plugindir, driver
and device mode2 falls back to the [lircd] section if not found in [mode2].
.IP "" 4
The location of this file can be changed using the -O/--options-file
command-line option or using the environment variable LIRC_OPTIONS_PATH.

.TP 4
.B ~/.cache/mode2.log
Debug log. Setting the XDG_CACHE_HOME environment variable relocates this
file to $XDG_CACHE_HOME/mode2.log. The amount of logging respects the
LIRC_LOGLEVEL environment variable, defaulting to the \fidebug\fR value
in lirc_options.conf.


.SH "SEE ALSO"
xmode2(1)
.br
lirc(4)

The documentation for
.B lirc
is maintained as html pages. They are located under html/ in the
documentation directory.


NEWS
0.9.4d 23/01/1e7
 * The debian packaging means that freebsd and hurd builds are
   regularly tested, improving build stability.
 * A multitude of build patches, mostly after building on Debian:
    + Cross-compilation FTBS fixed (#233)
    + Fix configuration of configure (#234)
    + Fix library specification in lirc.pc (#236)
    + use HAVE_UINPUT=1 to force building uinput code (#238).
    + lib: Only use local headers in lirc_private.h (#237).
    + Configure libudev (#242)
    + Various kfreebsd fixes
    + Include config.h in installed headers (#250).
    + Fix _XOPEN_SOURCE related FTBS on some platforms (#248)
 * lib: Fix some lircrcd-related bugs (#254)
 * systemd: Fix wrong path to irexec (#239)
 * systemd: Add missing lircd-setup.service to distribution (#243)
 * lircd: Fix --listen option parsing [#249]
 * lircd: Fix segfault in SET_INPUTLOG without arguments (#252)
 * lircd: Fix DRV_OPTION error message handling (#256).
 * ircat: Fix option parsing bug + empty conditions (#251)
 * xmode2: Fix handling of --driver option.

0.9.4c 22/10/16
 * New curl_poll files adds a new MIT license. Lirc now carries
   GPL, BSD and MIT licenses.
 * Several patches to make lirc work again on macos (#226, #227).
 * New tool irtext2udp generates udp plugin data (#197).
 * Fix FTBS errors on linux kernel >= 4.8.0 (#232).
 * plugins: girs: Add udp/receive support (#201).
 * mode2: Fix broken --raw option (#223).
 * lirc-make-devinput: Fix that two remotes has the same name (#222).
 * lircd: Warn for duplicate remote definitions (#222).
 * lircd: Retry temporary unavailable write sockets (#221).
 * contrib: Fix usb devices' acl permissions (rhbz #1364744).
 * lirc-setup: Fix crasher bugfix (Fedora abort trace).
 * contrib: disable udev /dev/uinput rule requiring lirc user.
 * plugins: audio_alsa: Fix byte truncating in 16-bit data (#218):
 * plugins: irtoy: improve error handling (#220).
 * Minor documentation fixes.

0.9.4b 9/8/16
  * lircd-uinput: A multitude of fixes (#213, #211, #177). Adds new options.
  * lirc_client:  Fix "Can't redirect irsend output" (#207).
  * lirc_client: read_string() not terminating buffer contents (#216)
  * Remove a single python2 dependency [4e3c97c].
  * irrecord: Fix under-sized buffer causing crash (#208).
  * systemd: Add hardening options comments to services (#204).
  * lircmd: Options, udev examples and manpage update.A
  * lib: Add missing Makefile.am dep (#210).
  * doc: Manpages and configuration guide fixes.
  * build: Don't define docdir - autoconf defines docdir (#215).
  * build: support runstatedir configure option (#212).

0.9.4a 22/6/16
  * Fix syntax errors in systemd unit files (rhbz: 1350750).
  * irrecord: Enhance "Don't use devinput" message (#180).
  * plugins/uirt2_raw: Reset device to UIR mode on deinit.
  * plugins/girs.c: Enable reception after sending.
  * lirc.org: Updates and bugfixes
  * lircd, liblirc_client: Fix freeaddrinfo handling (#195).
  * Don't build commandir unless we have usb.h (#191).
  * lirc_options.conf: debug => loglevel (#199).
  * lib: use proper linking method to avoid parallel build issue
  * tools: make_rel_symlink.py can also use python2
  * plugins/devinput: Make the list-devices support available.
  * configure.ac: Bugfix (BSD setting reported wrong).

0.9.4 22/5/16
  * Issues: https://sf.net/p/lirc/tickets/milestone/0.9.4
  * Digest of 0.9.4pre1, 0.9.4pre2

  * The caraca driver has been retired (still available in attic/).
  * The iguanair driver has been moved out of tree and is now
    maintained by Iguanaworks, Inc. See drivers/iguanair/README.
  * The irman driver has been moved out of tree. From 0.5.1 the
    libirman package contains the driver. See drivers/irman/README
  * The yard2 driver has been moved out of tree and is now maintained
    by the yard2 project. See driver/yard2/README.
  * The lirc.4 manpage (/dev/lirc* interface) is upstreamed to the
    linux kernel manpages project (#164). Old page lirc.4l still
    lingers in attic/ but is not installed.
  * Upstream selinux updated to match current lircd requirements.
    See: https://github.com/fedora-selinux/selinux-policy/pull/41

  * New upstream debian packaging with which Debian users can create
    .deb packages. Ubuntu packages are also available in PPAs. See
    README.md in the debian branch (#137).
  * New driver girs for the GIRS protocol, gives LIRC Arduino support.
    See new info in driver documentation.
  * New transmit-only FTDI driver with enhanced accuracy and hardware
    support. See ftdix driver documentation.
  * Support for non-modulated signals added (#132).
  * As shipped, lircd now uses the devinput driver and the new 'auto'
    device, should work for all remotes supported by kernel (#106).
  * Enhanced lirc-setup tool (#123, partly):
    - Add drvctl device enumeration support.
    - Integrated mode2 and irw tests of current configuration.
    - Documentation links.
    - More consistent data based on new database.
    - Multiple bugfixes.
  * Several patches to make mode2 work with all drivers(#123, partly)
  * Configuration database overhaul (#123, partly)
    - Database lives as yaml files in configs/.
    - Data is collected from drivers, remotes and static files.
    - Better ways for drivers to enumerate devices.
    - There is an unstable python database module used by
      all tools.
  * Several drivers including default and devinput has enhanced
    enumeration support, usable with mode2 -l and lirc-setup.
  * devinput.lircd.conf is generated at build time to match host kernel,
    new script lirc-make-devinput shipped to re-generate it (#153)
  * A new script lircd-setup runs as a one-shot service before lircd,
    performing setup commands as root (#121).
  * Several new examples in contrib/, notably lircrc templates for
    many popular LIRC applications and improved udev rules.
  * The default driver does not require root perms if the protocol
    is correct (#133).

tools/mode2.cpp
/****************************************************************************
** mode2.c *****************************************************************
****************************************************************************
*
* mode2 - shows the pulse/space length or decoded values of remote buttons.
*
*
* Copyright (C) 1998 Trent Piepho <xyzzy@u.washington.edu>
* Copyright (C) 1998 Christoph Bartelmus <lirc@bartelmus.de>
*
*/

/*
 * TODO: Close stuff (call curr_driver->close_func(), closing logs)
 * when a terminating signal arrives.
 */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#ifdef _CYGWIN_
#define __USE_LINUX_IOCTL_DEFS
#endif

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdarg.h>
#include <unistd.h>
#include <fcntl.h>
#include <getopt.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <limits.h>
#include <errno.h>
#include <glob.h>
#include <syslog.h>
#include <time.h>
#include <pwd.h>

#include "lirc_private.h"
#include "drv_enum.h"

static const logchannel_t logchannel = LOG_APP;

static const char* opt_device = NULL;
static const char* opt_driver = NULL;
static int opt_dmode = 0;
static int t_div = 500;
static unsigned int opt_gap = 10000;
static int opt_raw_access = 0;
static int opt_list_devices = 0;

static const char* const help =
	"Usage: mode2 [options]\n"
	"\t -d --device=device\tRead from given device\n"
	"\t -H --driver=driver\tUse given driver\n"
	"\t -U --plugindir=path\tLoad plugins from path\n"
	"\t -k --keep-root\t\tKeep root privileges\n"
	"\t -m --mode\t\tEnable column display mode\n"
	"\t -l --list-devices\tList available devices\n"
	"\t -r --raw\t\tAccess device directly without driver\n"
	"\t -g --gap=time\t\tTreat spaces longer than time as the gap\n"
	"\t -s --scope=time\t'Scope' like display with time us per char\n"
	"\t -A --driver-options=key:value[|key:value...]\n"
	"\t\t\t\tSet driver options\n"
	"\t -D --loglevel=level\t'error', 'info', 'notice',... or 3..10\n"
	"\t -O --options-file\tUse alternative lirc_options.conf file\n"
	"\t -h --help\t\tDisplay usage summary\n"
	"\t -v --version\t\tDisplay version\n";


static const struct option options[] = {
	{"help",           no_argument,       NULL, 'h'},
	{"debug",          required_argument, NULL, 'D'},
	{"version",        no_argument,       NULL, 'v'},
	{"device",         required_argument, NULL, 'd'},
	{"driver",         required_argument, NULL, 'H'},
	{"keep-root",      no_argument,       NULL, 'k'},
	{"list-devices",   no_argument,       NULL, 'l'},
	{"mode",           no_argument,       NULL, 'm'},
	{"raw",            no_argument,       NULL, 'r'},
	{"gap",            required_argument, NULL, 'g'},
	{"scope",          required_argument, NULL, 's'},
	{"plugindir",      required_argument, NULL, 'U'},
	{"driver-options", required_argument, NULL, 'A'},
	{0,	           0,		      0,    0  }
};

const char* const MSG_NO_GETMODE =
"Problems: this device is not a LIRC kernel device (it does not\n"
"support LIRC_GET_REC_MODE ioctl). This is not necessarily a\n"
"problem, but mode2 will not work.  If you are using the --raw\n"
"option you might try using without it and select a driver\n"
"instead. Otherwise, try using lircd + irw to view the decoded\n"
"data - this might very well work even if mode2 doesn't.\n";

const char* const USE_RAW_MSG =
"Please use the --raw option to access the device directly instead"
" through the abstraction layer\n";

static void add_defaults(void)
{
	const char* const device = options_getstring("lircd:device");
	const char* const driver = options_getstring("lircd:driver");
	const char* const plugindir = options_getstring("lircd:plugindir");
	const char* const lircd_level = options_getstring("lircd:debug");
	const char* const env_level = getenv("LIRC_LOGELEVL");
	char level[12];

	if (env_level != NULL) {
		if (string2loglevel(env_level) == LIRC_BADLEVEL) {
			fprintf(stderr,
			        "Bad debug level: %s in LIRC_LOGLEVEL\n",
			        env_level);
			exit(EXIT_FAILURE);
		}
		strncpy(level, env_level, sizeof(level) - 1);
	} else if (lircd_level != NULL) {
		strncpy(level, lircd_level, sizeof(level) - 1);
	} else {
		snprintf(level, sizeof(level), " %d", lirc_log_defaultlevel());
	}
	const char* const defaults[] = {
		"mode2:driver",     driver ? driver : "default",
		"mode2:lircdfile",  LIRCD,
		"lircd:logfile",    "syslog",
		"mode2:plugindir",  plugindir ? plugindir : PLUGINDIR,
		"lircd:configfile", LIRCDCFGFILE,
		"mode2:debug",	    level,
		"mode2:device",     device ? device : LIRC_DRIVER_DEVICE,
		(const char*)NULL,  (const char*)NULL
	};
	options_add_defaults(defaults);
}


static void parse_options(int argc, char** argv)
{
	int c;
	static const char* const optstring = "hvD:d:H:mklrg:s:U:A:";

	add_defaults();
	while ((c = getopt_long(argc, argv, optstring, options, NULL)) != -1) {
		switch (c) {
		case 'D':
			if (string2loglevel(optarg) == LIRC_BADLEVEL) {
				fprintf(stderr, "Bad debug level: %s\n", optarg);
				exit(EXIT_FAILURE);
			}
			options_set_opt("mode2:debug", optarg);
			break;
		case 'h':
			puts(help);
			exit(EXIT_SUCCESS);
		case 'H':
			options_set_opt("mode2:driver", optarg);
			break;
		case 'v':
			printf("%s %s\n", "mode2 ", VERSION);
			exit(EXIT_SUCCESS);
		case 'U':
			options_set_opt("mode2:plugindir", optarg);
			break;
		case 'k':
			unsetenv("SUDO_USER");
			break;
		case 'l':
			opt_list_devices = 1;
			break;
		case 'd':
			options_set_opt("mode2:device", optarg);
			break;
		case 's':
			opt_dmode = 2;
			t_div = atoi(optarg);
			break;
		case 'm':
			opt_dmode = 1;
			break;
		case 'r':
			opt_raw_access = 1;
			break;
		case 'g':
			opt_gap = atoi(optarg);
			break;
		case 'A':
			options_set_opt("lircd:driver-options", optarg);
			break;
		default:
			fprintf(stderr, "Usage: mode2 [options]\n");
			exit(EXIT_FAILURE);
		}
	}
	if (optind < argc) {
		fputs("Too many arguments\n", stderr);
		exit(EXIT_FAILURE);
	}
	opt_device = options_getstring("mode2:device");
	if (opt_raw_access && opt_device == NULL) {
		fprintf(stderr, "The --raw option requires a --device\n");
		exit(EXIT_FAILURE);
	}
	options_set_opt("lircd:plugindir",
			options_getstring("mode2:plugindir"));
	opt_driver = options_getstring("mode2:driver");
	if (hw_choose_driver(opt_driver) != 0) {
		fprintf(stderr, "Driver `%s' not found.", opt_driver);
		fputs(" (Missing -U/--plugins option?)\n", stderr);
		fputs("\nAvailable drivers:\n", stderr);
		hw_print_drivers(stderr);
		exit(EXIT_FAILURE);
	}
}


/** Open device using curr_driver->open_func() and curr_driver->init_func().*/
int open_device(int opt_raw_access, const char* device)
{
	struct stat s;

	uint32_t mode;
	int fd;
	const char* opt;

	log_info("Opening device: %s", device);
	if (opt_raw_access) {
		fd = open(device, O_RDONLY);
		if (fd == -1) {
			device = device ? device : "(Null)";
			perrorf("Error while opening device: %s", device);
			exit(EXIT_FAILURE);
		}
		;
		if ((fstat(fd, &s) != -1) && (S_ISFIFO(s.st_mode))) {
			/* can't do ioctls on a pipe */
		} else if ((fstat(fd, &s) != -1) && (!S_ISCHR(s.st_mode))) {
			fprintf(stderr, "%s is not a character device\n",
				device);
			fputs("Use the -d option to specify device\n",
			      stderr);
			close(fd);
			exit(EXIT_FAILURE);
		} else if (ioctl(fd, LIRC_GET_REC_MODE, &mode) == -1) {
			fputs(MSG_NO_GETMODE, stderr);
			close(fd);
			exit(EXIT_FAILURE);
		}
	} else {
		curr_driver->open_func(device);
		opt = options_getstring("lircd:driver-options");
		if (opt != NULL)
			if (drv_handle_options(opt) != 0) {
				fprintf(stderr,
				"Cannot set driver (%s) options (%s)\n",
				curr_driver->name, opt);
				exit(EXIT_FAILURE);
			}
		if (!curr_driver->init_func || !curr_driver->init_func()) {
			fprintf(stderr, "Cannot initiate device %s\n",
				curr_driver->device);
			exit(EXIT_FAILURE);
		}
		fprintf(stderr, "Trying device: %s\n", curr_driver->device);
		fd = curr_driver->fd;   /* please compiler */
		mode = curr_driver->rec_mode;
		if (mode != LIRC_MODE_MODE2) {
			if (strcmp(curr_driver->name, "default") == 0) {
				fputs(USE_RAW_MSG, stderr);
				exit(EXIT_FAILURE);
			} else if (mode != LIRC_MODE_LIRCCODE) {
				fputs("Internal error: bad receive mode\n",
				      stderr);
				exit(EXIT_FAILURE);
			}
		}
	}
	if (opt_device && strcmp(opt_device, LIRCD) == 0) {
		fputs("Refusing to connect to lircd socket\n", stderr);
		exit(EXIT_FAILURE);
	}
	fprintf(stderr, "Using device: %s\n", curr_driver->device);
	return fd;
}


/** Get the codelength (bits per decoded value) for lirccode data. */
unsigned int get_codelength(int fd, int use_raw_access)
{
	unsigned int code_length;
	int r = 1;

	if (use_raw_access) {
		if (ioctl(fd, LIRC_GET_LENGTH, &code_length) == -1) {
			perror("Could not get code length");
			close(fd);
			exit(EXIT_FAILURE);
		}
		return code_length;
	}
	if (curr_driver->drvctl_func) {
		r = curr_driver->drvctl_func(DRVCTL_GET_RAW_CODELENGTH,
					     &code_length);
	}
	if (r != 0)
		code_length = curr_driver->code_length;
	return code_length;
}


/** Print mode2 data as pulse/space durations, one per line. */
void print_mode2_data(unsigned int data)
{
	static int bitno = 1;

	switch (opt_dmode) {
	case 0:
		printf("%s %u\n", (
			       data & PULSE_BIT) ? "pulse" : "space",
		       (uint32_t)(data & PULSE_MASK));
		break;
	case 1: {
		/* print output like irrecord raw config file data */
		printf(" %8u", (uint32_t)data & PULSE_MASK);
		++bitno;
		if (data & PULSE_BIT) {
			if ((bitno & 1) == 0)
				/* not in expected order */
				fputs("-pulse", stdout);
		} else {
			if (bitno & 1)
				/* not in expected order */
				fputs("-space", stdout);
			if (((data & PULSE_MASK) > opt_gap) || (bitno >= 6)) {
				/* real long space or more
				 * than 6 codes, start new line */
				puts("");
				if ((data & PULSE_MASK) > opt_gap)
					puts("");
				bitno = 0;
			}
		}
		break;
	}
	case 2:
		if ((data & PULSE_MASK) > opt_gap) {
			fputs("_\n\n_", stdout);
		} else {
			printf("%.*s",
			       ((data & PULSE_MASK) + t_div / 2) / t_div,
			       (data & PULSE_BIT) ?
			       "------------" : "____________");
		}
		break;
	}
	fflush(stdout);
}


/** Print lirccode data as a decoded value (an integer) per line. */
void print_lirccode_data(char* buffer, size_t count)
{
	size_t i;

	fputs("code: 0x", stdout);
	for (i = 0; i < count; i++)
		printf("%02x", (unsigned char)buffer[i]);
	puts("");
	fflush(stdout);
}


/**
 * Process next button press and print a dump, return boolean OK/FAIL.
 */
int next_press(int fd, int mode, int bytes)
{
	char buffspace[bytes];
	int r;
        union {
		char* buffer;
		lirc_t data;
	} input;

	input.buffer = buffspace;
	if (opt_raw_access || mode != LIRC_MODE_MODE2) {
		r = read(fd, input.buffer, bytes);
		if (r == -1)
			perrorf("read() error on %s", opt_device);
		else if (r != (int)bytes)
			fprintf(stderr, "Partial read %d bytes on %s",
				r, opt_device);
		if (r != (int)bytes)
			return 0;
		if (mode == LIRC_MODE_MODE2)
			print_mode2_data(input.data);
		else
			print_lirccode_data(input.buffer, bytes);
	} else {
		input.data = curr_driver->readdata(0);
		if (input.data == 0) {
			fputs("readdata() failed\n", stderr);
			return 0;
		}
		print_mode2_data(input.data);
	}
	return 1;
}


static void list_devices(void)
{
	glob_t glob;
	unsigned i;
	int r;

	r = curr_driver->drvctl_func ? 0 : DRV_ERR_NOT_IMPLEMENTED;
	if (r == 0) {
		r = curr_driver->drvctl_func(DRVCTL_GET_DEVICES, &glob);
	}
	switch (r) {
		case DRV_ERR_ENUM_EMPTY:
		break;
	case DRV_ERR_NOT_IMPLEMENTED:
		fputs("Device enumeration is not supported by this driver\n",
			stderr);
		exit(1);
	case 0:
		for (i = 0; i < glob.gl_pathc; i += 1)
			puts(glob.gl_pathv[i]);
		curr_driver->drvctl_func(DRVCTL_FREE_DEVICES, &glob);
		break;
	default:
		fputs("Error running enumerate command", stderr);
		exit(1);
	}
}


int main(int argc, char** argv)
{
	int fd;
	uint32_t mode;
	size_t bytes = sizeof(lirc_t);
	char logpath[256];
	/**
	 * Was hard coded to 50000 but this is too long, the shortest gap in the
	 * supplied .conf files is 10826, the longest space defined for any one,
	 * zero or header is 7590
	 */
	uint32_t code_length;
	const loglevel_t level = options_get_app_loglevel("mode2");

	hw_choose_driver(NULL);
	options_load(argc, argv, NULL, parse_options);

	lirc_log_get_clientlog("mode2", logpath, sizeof(logpath));
	(void)unlink(logpath);
	lirc_log_set_file(logpath);
	lirc_log_open("mode2", 1, level);
	if (opt_list_devices) {
		list_devices();
		return 0;
	}
        if (opt_raw_access) {
		printf("Using raw access on device %s\n", opt_device);
	} else {
		printf("Using driver %s on device %s\n",
		       options_getstring("mode2:driver"),
		       opt_device);
	}
	fd = open_device(opt_raw_access, opt_device);
	if (geteuid() == 0)
		drop_root_cli(setuid);

	mode = curr_driver->rec_mode;
	if (mode == LIRC_MODE_LIRCCODE) {
		code_length = get_codelength(fd, opt_raw_access);
		bytes = (code_length + CHAR_BIT - 1) / CHAR_BIT;
	}
	while (next_press(fd, mode, bytes))
		;
	return EXIT_SUCCESS;
}
